PC/pyconfig.h

```diff
@@ -388,6 +399,11 @@ Py_NO_ENABLE_SHARED to find out.  Also support MS_NO_COREDLL for b/w compat */
 /* Define to 1 if you have the <conio.h> header file. */
 #define HAVE_CONIO_H 1
 
+/* Define to 1 if you support windows console io */
+#if defined(MS_WINDOWS_APP) || defined(MS_WINDOWS_SYSTEM)
```

Comment:
This is an example of a condition that makes me question things. _Should_ we have `DESKTOP` in here as well?

Personally, I'd include it even if it's redundant because it makes things clearer. (At one point in history, apps did not support console IO. It was brought back later. Maybe I'm one of the only two people who ever made Python run under those circumstances? And so everyone else will make different assumptions)

Reply:
Yeah, please do. I think otherwise we'll end up needing comments everywhere saying "reminder: app is a superset of desktop", so we may as well just explicitly mention desktop.

Reply:
I updated them

Reply:
> This is an example of a condition that makes me question things. Should we have `DESKTOP` in here as well?

It would be nice to know what the minimum supported SDK version will be for building 3.12. Currently "PCbuild/readme.txt" requires Visual Studio 2017. Will this get bumped up to Visual Studio 2019?

On a related note, "PC/pyconfig.h" still targets Windows 8. Will `Py_NTDDI` get changed to `NTDDI_WIN10`? I suppose given LTSC releases of Windows 10, Python 3.12 has to support running on all versions back to 10.0.10240.

> I think otherwise we'll end up needing comments everywhere saying "reminder: app is a superset of desktop"

The `APP` API partition is the intersection of the `DESKTOP`, `PC_APP`, and `PHONE_APP` API partitions. A lot of the Windows API is declared for the `DESKTOP` partition but isn't available in the `APP` partition.

When we require the `GAMES` API partition, will we need to redundantly include the `DESKTOP` API partition? Doing so isn't a technical problem, at least not currently, because the `DESKTOP` partition only includes the desktop family. Still, it doesn't feel right to me. Similarly, by that logic, since the `SYSTEM` API partition includes the system and server families, requiring the `SYSTEM` API partition would need to also require the `SERVER` API partition (defined in "winpackagefamily.h") in order to clarify the included families. This just feels like using API families instead of API partitions.

Reply:
> It would be nice to know what the minimum supported SDK version will be for building 3.12

For example, version 10.0.19041 of the SDK is the first version that makes "memoryapi.h" available for `GAMES`, such as for `MapViewOfFile()`. It's also the first version that makes `OpenFileMappingA()` and `CreateFileMappingA()` available for `GAMES` in "winbase.h". Thus building `mmap` for Xbox requires at least version 10.0.19041 of the Windows SDK.

Reply:
> This feels like using API families instead of API partitions

Yes, which is why I said we should go with families, because it makes the code easier to reason about when you're reading it in isolation. I suspect you (@eryksun) are _literally_ the only person here who'll casually remember that `app` implies `desktop`, because I certainly won't! And I shouldn't have to look up reference manuals just to read the code - it should be obvious enough without obscure context.

> It would be nice to know what the minimum supported SDK version will be for building 3.12

This is not something we've ever really dealt with before. In general, you _should_ use the latest available SDK, because the macros should handle API availability, and other fixes we get those as well. That said, I know we have some contributors who are still using the earliest Windows 10 SDK release and will not/have not updated it.

I think the easiest "rule" to bring in is to require as much support as possible for CI builds (debug, desktop partition, using prebuilt dependencies) to support as many contributors as possible. Then it's okay to require a newer SDK for special builds and release builds. If there's a need to enforce that, we should be able to detect it in headers or python.props.

> On a related note, "PC/pyconfig.h" still targets Windows 8. Will Py_NTDDI get changed to NTDDI_WIN10?

I thought we had one more release including 8.1? Or was 3.11 that release and I'm mixing up timings? Either way, we can probably update it, but until we have a need, there's no hurry.

Reply:
Core API sounds like it is a subset supported by all targets, so I am not a big fan of it.

Reply:
Agreed, which then means we either use `APP`, which sounds like it _isn't_ supported by desktop applications, or we just specify `DESKTOP || APP || SERVER` when we want it for all of those and then `GAMES` more obviously has no overlap.

Reply:
The application API family names are `WINDOWS_FAMILY_DESKTOP_APP`, `WINDOWS_FAMILY_PC_APP`, and `WINDOWS_FAMILY_PHONE_APP`. It seems clear to me that `WINDOWS_PARTITION_APP` is defined when targeting any of the `APP` API families.
 
> I think the easiest "rule" to bring in is to require as much support as possible for CI builds (debug, desktop partition, using prebuilt dependencies) to support as many contributors as possible.

Maybe we should check whether an `NTDDI_*` macro is defined. For example, using SDK version 10.0.18362 and earlier, `mmap` should only be built for `DESKTOP`. When using 10.0.19041 (`NTDDI_WIN10_VB`, "sdkddkver.h") or later, it's available for `DESKTOP` and `GAMES`.

```c
#if !defined(MS_WINDOWS) || defined(MS_WINDOWS_DESKTOP) || (defined(NTDDI_WIN10_VB) && defined(MS_WINDOWS_GAMES))
```

> I thought we had one more release including 8.1? Or was 3.11 that release and I'm mixing up timings?

Extended support for Windows 8.1 ended on 2023-01-10. Server 2012 R2 (NT 6.3) is supported until [2023-10-10](https://learn.microsoft.com/en-us/windows/release-health/status-windows-8.1-and-windows-server-2012-r2). The last time we discussed this, you said that 3.12 would not support Server 2012 R2. The planned released date for 3.12 is 2023-10-02.

Based on my reading of PEP 11, LTSC releases of Windows 10 have to be supported as long as Microsoft supports them. For example, Windows 10 Enterprise LTSC 10.0.10240 has extended support until [2025-10-14](https://learn.microsoft.com/en-us/windows/release-health/release-information).

Reply:
> It seems clear to me that `WINDOWS_PARTITION_APP` is defined when targeting any of the APP API families.

Yes, _with the additional context that you know the families that are part of that partition_. So we need to add a comment next to the `ifdef` letting the reader know which families are included... or just include the redundant partition in the check. The aim is to help the ignorant person guess correctly, whether with an explanatory comment or clear code. Nothing is gained by expecting people to have context that isn't immediately obvious or general[^1] knowledge

[^1]: "General" for someone who's reading Win32 API calls written in C.

> The last time we discussed this, you said that 3.12 would not support Server 2012 R2.

That guy clearly knew what he was talking about ;-) So we're up to 10.0.10240, which really just means we need to dynamically load anything since then, and be careful not to use constants in arguments if they were defined later. I think we already required using the Windows 10 SDK though, so nothing changes yet.

> Maybe we should check whether an `NTDDI_*` macro is defined. For example, using SDK version 10.0.18362 and earlier, mmap should only be built for DESKTOP. 

I was more thinking we could detect it and `#error`  if we require a newer version for certain families.

For `CreateFileMappingA` I've done as much archeology as I can (involved tracking down our archive server), and it seems that the API entry point was omitted from `xgaming.lib` to try and deprecate it, and it was brought back because people needed it. But that's just the "umbrella" library that connect up all the imports through API sets, and it apparently never moved location, so if you build with a newer GDK it still ought to work on older Xbox. I don't have any way to confirm that myself though.

I'm inclined to say `MS_WINDOWS_GAMES` requires a newer WinSDK to avoid that issue, and if you don't have one then you'll get build errors. But @maxbachmann is the one trying to build it, so if updating the GDK is impossible for some reason I'm not aware of, I'd love to hear it. (Though it's starting to look like the easiest fix is to finally update `mmap` to use the `*W` APIs...)

Reply:
We are already using  `10.0.19041`. In addition the whole engine is only used in house, so upgrading is generally not to much pain. Still I guess it should not be an ungodly amount of work to update `mmap` to use the `*W` APIs either.

Reply:
The compatibility concerns about changing what constitutes a valid tagname are the biggest risk with changing `mmap`. I expect they're almost certainly non-existent, but we need to at least consider it. We should probably do that before 3.12 releases so we can deprecate it if that's what we need to do.

Reply:
> The compatibility concerns about changing what constitutes a valid tagname

@zooba, currently `tagname` gets encoded as UTF-8. The Windows API best-fit decodes it using the ANSI code page of the current process. The latter can be UTF-8 nowadays, but it's still much more likely to be a legacy code page. It's possible that existing scripts have hard-coded some nonsense mojibake name for IPC with another application, especially if the other application creates the section object.

Reply:
If the section is shared between Python 3.12 and an older version of Python, in 3.12 a non-ASCII name will need to be passed as `tagname=name.encode('utf-8').decode('mbcs', 'replace')`.